transhumandesign/kag-base
# [added] When a player dies, blood will get drawn on walls temporarily

## Status

- **IN DEVELOPMENT** 
- There are unfixed issues, see below.

## Description

`[added] when a player character dies, blood stains will be drawn on nearby tiles`

Video: https://www.youtube.com/watch?v=p3SlVDi8Wog

This PR adds a sprite png file `BloodStains.png` and a script `BloodStains.as` which is added to gamemode files for Sandbox, CTF, small CTF, TDM and TTH.

When a player character is killed by suicide, sword, bomb, explosion, keg, mine, mine_special or saw, the script will find walls near the killed player and draw blood on them that will last 30 seconds or until the tile collapses or until the tile is destroyed.

For now, I didn't use `v_fastrender` to make it only show when `faster graphics` is off. Perhaps blood stains should have its own option for turning it on and off in the menu. Since people might like fastrender to turn off the background but may also wish for blood stains to be shown.

This was tested in offline and online and it works as intended for the most part. There are open issues which are listed below. I don't know if I will get to fixing these.

## Remaining issues

The function `GUI::DrawIcon()` is used which relies on `driver.getResolutionScaleFactor();` which is bugged right now (related issue: https://github.com/transhumandesign/kag-base/issues/1711 ). Everything looks normal on 16:9 resolutions but it looks off when using 4:3 resolutions. It was said in discord this is going to get fixed in the next build. I didn't find another way to draw my sprites on screen.

Another issue which I didn't find a fix for is that when `/rcon rebuild` is used, the blood stains script will not work anymore until restarting the game again.

On a rare occasion a wrong frame is used, i.e. an up-facing blood stain when it should be a right-facing one.

The drawn blood does not account for shading when being underground or at the border of the map. As I said, I didn't find another way to draw my sprites so I had to rely on what `GUI::DrawIcon()` can offer.

It may be possible to sum up these four in a single function:

```
checkGround(this, victimPosBottom, map);
	checkLeftWall(this, victimPosLeft, map);
	checkRightWall(this, victimPosRight, map);
	checkCeiling(this, victimPosTop, map);
```

## Steps to Test or Reproduce

Go to online or offline.
Use `AddBot("");` and kill the bot near a wall or use the suicide button in menu.
Notice blood getting added to the wall.
Try to wait 30 sec or destroy or collapse the tile that has the blood on it.
Notice the blood will be removed.
